src/main/java/com/onwelo/tinvalidator/validator/ESValidator.java

```diff
@@ -0,0 +1,101 @@
+
+    private static final Pattern PATTERN = Pattern.compile("[0-9A-Z]\\d{7}[0-9A-Z]");
+
+    @Override
+    public Pattern getPattern() {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean computeControlSum(String tin) {
+        try {
+            Pattern assertFirstCharactersAllowedIfLastIsALetter = Pattern.compile("[A-HK-NPQ-SW-Z0-9]");
+            Pattern assertFirstCharactersAllowedIfLastIsADigit = Pattern.compile("[A-HJUV]");
+            Pattern firstCharactersForNonNationalJuridicalEntities = Pattern.compile("[A-HNPQ-SW]");
+            Pattern firstCharactersForPhysicalPersons = Pattern.compile("[K-MX-Z0-9]");
+
+            String lastCharacter = tin.substring(tin.length() - 1);
+            String firstCharacter = tin.substring(0, 1);
+            boolean isLastCharacterNumeric = isNumeric(lastCharacter);
+
+            assert (isLastCharacterNumeric && assertFirstCharactersAllowedIfLastIsADigit.matcher(firstCharacter).matches())
+                    || (!isLastCharacterNumeric && assertFirstCharactersAllowedIfLastIsALetter.matcher(firstCharacter).matches());
+
+            if (!isLastCharacterNumeric && firstCharactersForNonNationalJuridicalEntities.matcher(firstCharacter).matches()) {
+                return computeControlSumForNonNationalJuridicalEntities(tin);
+            } else if (!isLastCharacterNumeric && firstCharactersForPhysicalPersons.matcher(firstCharacter).matches()) {
+                return computeControlSumForPhysicalPersons(tin);
+            } else {
+                return computeControlSumForNationalJuridicalEntities(tin);
+            }
+        } catch (NumberFormatException | AssertionError e) {
+            return false;
+        }
+    }
+
+    private boolean computeControlSumForNonNationalJuridicalEntities(String tin) {
+        final String[] checkCharacters = new String[]{"A", "B", "C", "D", "E", "F", "G", "H", "I", "J"};
+        String lastCharacter = tin.substring(tin.length() - 1);
+
+        return checkCharacters[calculateRForJuridicalEntities(tin) - 1].equals(lastCharacter);
+    }
+
+
+    private boolean computeControlSumForNationalJuridicalEntities(String tin) {
+        String lastCharacter = tin.substring(tin.length() - 1);
+
+        return Integer.parseInt(lastCharacter) == calculateRForJuridicalEntities(tin) % 10;
+    }
+
+    private boolean computeControlSumForPhysicalPersons(String tin) {
+        final String[] checkCharacters = new String[]{"T", "R", "W", "A", "G", "M", "Y", "F", "P", "D", "X", "B", "N",
+                "J", "Z", "S", "Q", "V", "H", "L", "C", "K", "E"};
+        String lastCharacter = tin.substring(tin.length() - 1);
+        String firstCharacter = tin.substring(0, 1);
+        int c1 = 0;
+
+        if (firstCharacter.equals("Y")) {
+            c1 = 1;
+        } else if (firstCharacter.equals("Z")) {
+            c1 = 2;
+        }
+
+        int startIndex = 1;
+        if (!isNumeric(firstCharacter) && c1 > 0) {
```

Comment:
Skąd jest ten warunek `c1 > 0`?

Reply:
to jest mój warunek, sprawdzam czy wcześniej była jakaś zmiana w wartości tej zmiennej żeby wiedzieć czy mam ją dokleić do nipu